### ISSUE TITLE
Disallow not null assertion (!!)

### ISSUE THREAD

**Issue by alyssa-glean** [detekt/detekt#4832]
## Expected Behavior of the rule
Quite simply, a rule which disallows any use of the [not null operator](https://kotlinlang.org/docs/null-safety.html#the-operator) `!!`. Whilst there *are* ways to use it safely, you can always avoid it by instead using optional chaining, the elvis operator, a `?.let{}` and so on. 

The rule would of course default to off, and for my own use case I imagine turning it on for production code only (`!!` is a handy convenience when writing tests sometimes). 

Noncompliant:

```kt
var someVar: SomeClass? = null

if (someVar != null) {
    someVar!!.doSomething() // !! is still needed despite the null check as the property is mutable it can't be smart-casted to non-null
}
```

Compliant:

```kt
var someVar: SomeClass? = null

someVar?.doSomething()
```

or

```kt
var someVar: SomeClass? = null

someVar?.let { someVar ->
    someVar.doSomething()
    someVar.doSomethingElse()
    ...
}
```

or

```kt
var someVar: SomeClass? = null

val currentValue = someVar ?: throw MyCustomException()
currentValue.doSomething()
```

## Context

One of the great things about Kotlin is the null-safety that it provides, but use of the not null operator reintroduces the potential for NullPointerExceptions. Being able to lint against it ensures nullable values are always handled appropriately. 

I'm happy to try and pick this up as a contributor if there's support for it :smile:

**Comment by cortinico** [comment]
We do have `UnnecessaryNotNullOperator` which does something similar. Though we don't have a rule which blindly flags usages of `!!`.

**Comment by BraisGabin** [comment]
I think that `UnsafeCallOnNullableType` does exactly that. Right?

**Comment by schalkms** [comment]
What would be the difference to the `UnsafeCallOnNullableType` rule?

**Comment by alyssaburlton** [comment]
Ah yes - I think `UnnecessaryNotNullOperator` and `UnsafeCallOnNullableType` in combination cover what I'm after. I haven't gotten around to exploring the rules that require type resolution yet which is why I hadn't noticed these.

What I was proposing was basically a combination of the two that wouldn't need type checking at all - just ban all `!!` without caring whether it's "unsafe" or "unnecessary" (both are bad). But the existing rules are more flexible, and dropping the need for type resolution feels like a fringe benefit so probably not worth changing what you already have. 

I'll just get on running detekt with type resolution to get what I'm after in my projects :grin:
